### PR TITLE
feat(iot-dev, deps): Add proxy support for MQTT_WS

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Adding proxy support to MQTT_WS.

Paho has published version 1.2.2 which fits our needs for proxy support without breaking any existing features. Version 1.2.0 had a bug that prevented setting the sslsocketfactory from working correctly, and version 1.2.1 had a bug that broke all websocket communication.

Intentionally avoiding using the Socket constructor that takes a proxy as it does not support HTTP proxies on Android or in Java 1.7 or lower. Had to manually create a SSLProxiedSocket to fit this requirement.

```
Socket proxySocket = new Socket(proxy);
```

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/Socks5SocketFactory.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        return createSocketImpl(new InetSocketAddress(host, port));
+    }
+
+    @Override
+    public Socket createSocket(final String host, final int port, final InetAddress localHost, final int localPort) throws IOException
+    {
+        return createSocketImpl(new InetSocketAddress(host, port), localHost, localPort);
+    }
+
+    private Socket createSocketImpl(final InetSocketAddress host, final InetAddress localHost, final int localPort) throws IOException
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/Socks5SocketFactory.java]: It too much overhead. Please remove `createSocketImpl` but call  `createSocket`.

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/Socks5SocketFactory.java]: Sure thing, I've simplified this a bit

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+     * @param port The destination port the proxy will forward communication to
+     * @throws IOException If unable to read or send to the HTTP proxy
+     */
+    private void doTunnelHandshake(Socket tunnel, String host, int port) throws IOException
+    {
+        Charset byteEncoding = StandardCharsets.UTF_8;
+        OutputStream out = tunnel.getOutputStream();
+        String msg = "CONNECT " + host + ":" + port + " " + HTTP_VERSION + "\n"
+                + "Proxy-Connection: keep-alive\n"
+                + "Host: " + host + "\n"
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java]: Missing `port` in host header

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                + "User-Agent: " + userAgentString + "\n"
+                + "Content-type: text/plain; charset=" + byteEncoding + "\n"
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java]: Please remove other headers in `CONNECT` request. Following is an example:
```
CONNECT server.example.com:80 HTTP/1.1
Host: server.example.com:80
Proxy-Authorization: basic aGVsbG86d29ybGQ=
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                + "Accept-Charset: " + byteEncoding
+                + "\r\n\r\n";
+
+        byte[] b = msg.getBytes(byteEncoding);
+
+        out.write(b);
+        out.flush();
+
+        byte[] reply = new byte[HTTP_PROXY_CONNECT_RESPONSE_MAX_SIZE];
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java]: Use `ByteArrayOutputStream` please.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        while (newlinesSeen < 2)
+        {
+            int i = in.read();
+            if (i < 0)
+            {
+                throw new IOException("Unexpected EOF from proxy");
+            }
+
+            if (i == '\n')
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java]: Offline discussion. It's not follow HTTP protocol.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+     * @param tunnel The socket to communicate to the HTTP proxy through
+     * @param host The destination host the proxy will forward communication to
+     * @param port The destination port the proxy will forward communication to
+     * @throws IOException If unable to read or send to the HTTP proxy
+     */
+    private void doTunnelHandshake(Socket tunnel, String host, int port) throws IOException
+    {
+        Charset byteEncoding = StandardCharsets.UTF_8;
+        OutputStream out = tunnel.getOutputStream();
+        String msg = "CONNECT " + host + ":" + port + " " + HTTP_VERSION + "\n"
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/ProxiedSSLSocket.java]: Please use string format

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/jproxy/impl/ProxyNegotiationHandler.java [hunk 7]
```diff
@@ -175,6 +174,8 @@ private void copyBytesBetweenClientAndService(final ReadWriteState readWriteStat
             final Scanner requestScanner = new Scanner(request);
             final String firstLine = requestScanner.nextLine();
 
+            System.out.println(request);
```
Comment (davilu) [inline iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/jproxy/impl/ProxyNegotiationHandler.java]: Please use `log`

Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/jproxy/impl/ProxyNegotiationHandler.java]: I'll remove this print, it was added by mistake

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+    private interface ExcludedListMethods
+    {
+        void connect(SocketAddress socketAddress, int timeout) throws IOException;
+        void connect(SocketAddress socketAddress) throws IOException;
+        void close() throws IOException;
+    }
+
+    @AllArgsConstructor
+    class HttpHeaderLineInputStream
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: It's not following HTTP protocol.
```
    class HttpHeaderLineInputStream
    {
        private final InputStream inputStream;
        private final Charset charset;
        private final ByteArrayOutputStream httpLineOutputStream;
        private boolean done;
        private boolean firstLine;
        HttpHeaderLineInputStream(InputStream inputStream, Charset charset) {
            this.inputStream = inputStream;
            this.charset = charset;
            this.httpLineOutputStream = new ByteArrayOutputStream();
            this.firstLine = true;
        }

        String readUntilHttpNewline() throws IOException
        {
            if (done) {
                return "";
            }
            
            int crlf = 0;
            while (crlf < 2)
            {
                int i = inputStream.read();
                if (i == -1)
                {
                    inputStream.close();
                    throw new IOException("Unexpected EOF from proxy");
                }
                httpLineOutputStream.write((byte) i);
                if ((i == '\r' && crlf == 0) || (i == '\n' && crlf == 1))
                {
                    crlf++;
                }
            }
            String httpHeaderLine = httpLineOutputStream.toString(charset.name());
            httpLineOutputStream.reset();
            if (httpHeaderLine.length() == 0) {
                if (firstLine) {
                    return readUntilHttpNewline();
                } else {
                    done = true;
                }
            } else {
                firstLine = false;
            }
            return httpHeaderLine;
        }
    }
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/HttpProxySocketFactory.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    private final ProxySettings proxySettings;
+
+    @Override
+    public Socket createSocket() throws IOException
+    {
+        Socket proxySocket = new Socket(proxySettings.getHostname(), proxySettings.getPort());
+        return new ProxiedSSLSocket(delegate, proxySocket, proxySettings.getUsername(), proxySettings.getPassword());
+    }
+
+    private interface ExcludedListMethods
```
[the remark this answers is not in the document]
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/HttpProxySocketFactory.java]: I'll call these something like SSLSocketFactoryNonDelegatedFunctions instead, sure

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+        String proxyConnectMessage;
+        if (this.proxyUsername != null && this.proxyPassword != null)
+        {
+            String base64EncodedCredentials = Base64.encodeBase64String(String.format("%s:%s", this.proxyUsername, new String(this.proxyPassword)).getBytes(byteEncoding));
+            proxyConnectMessage = String.format("CONNECT %s %s\r\nHost: %s\r\nProxy-Authorization: Basic %s\r\n\r\n", hostWithPort, HTTP_VERSION_1_1, hostWithPort, base64EncodedCredentials);
+        }
+        else
+        {
+            proxyConnectMessage = String.format("CONNECT %s %s\r\nHost: %s\r\n\r\n\r\n", hostWithPort, HTTP_VERSION_1_1, hostWithPort);
```
[the remark this answers is not in the document]
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: good catch

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        if (connectResponseStatusCode <= 199 || connectResponseStatusCode >= 300)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to return status code 2XX but status code was %d", host, port, connectResponseStatusCode));
+        }
+
+        log.trace("HTTP proxy responded to connect request with status {}, so the proxy connect was successful", connectResponseStatusCode);
+    }
+
+    private interface ExcludedListMethods
```
[the remark this answers is not in the document]
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: sure

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java (hunk 8 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: I'll have this read until /r/n/r/n instead

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+                int i = inputStream.read();
+                if (i == -1)
+                {
+                    inputStream.close();
+                    throw new IOException("Unexpected EOF from proxy");
+                }
+
+                httpLineOutputStream.write((byte) i);
+
+                if (i == '\n')
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: What if `\r\n\r\r\r\n`?

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: Okay, I've added logic to catch that, and also to catch ```\r\n\n\r\n```

Comment (davilu) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: How about:
```
if (i == '\n' && previousCharacter == '\r') {
  consecutiveHttpNewlines++;
} else if (i != '\r'||previousCharacter == '\n' ) {
  consecutiveHttpNewlines=0;
}
```

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: Now, I'll just keep a char[] of the last 4 read chars, and loop until they match ```\r\n\r\n```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+        String readHttpConnectResponse() throws IOException
+        {
+            if (alreadyRead)
+            {
+                throw new IOException("Http connect response has already been read");
+            }
+
+            ByteArrayOutputStream httpLineOutputStream = new ByteArrayOutputStream();
+            char[] mostRecentFourCharacters = new char[4];
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: `LinkedList` will be the best fit if you want to keep 4 chars.

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: Okay, I've switched to using a linked list instead

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+                if (i == -1)
+                {
+                    inputStream.close();
+                    throw new IOException("Unexpected EOF from proxy");
+                }
+
+                httpLineOutputStream.write((byte) i);
+
+                //[0] was most recent. When another character is read, push it to [1], then [2], etc.
+                mostRecentFourCharacters.removeLast();
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]:
```
mostRecentFourCharacters.poll();
mostRecentFourCharacters.push((char) i);
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+            ByteArrayOutputStream httpLineOutputStream = new ByteArrayOutputStream();
+            LinkedList<Character> mostRecentFourCharacters = new LinkedList<>();
+            mostRecentFourCharacters.push(' ');
+            mostRecentFourCharacters.push(' ');
+            mostRecentFourCharacters.push(' ');
+            mostRecentFourCharacters.push(' ');
+
+            //until the 4 most recently read characters were \r\n\r\n
+            while (!(mostRecentFourCharacters.get(0) == '\n' && mostRecentFourCharacters.get(1) == '\r' && mostRecentFourCharacters.get(2) == '\n' && mostRecentFourCharacters.get(3) == '\r'))
```
Comment (davilu) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: No need to init the list
```
while (!isCRLF(mostRecentFourCharacters))

// function
boolean isCRLF(List<Charater> list) {
  if (list.size() < 4) { 
    return false;
  }
  return list.get(0) == '\r' && list.get(1) == '\n' && list.get(2) == '\r' && list.get(3) == '\n';
}
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java (hunk 14 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: okay

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java (hunk 15 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java]: okay

**Pull request merged**